### PR TITLE
Home Assistant Custom Component Errata

### PR DESCRIPTION
**Pull request by mweinelt**
## Description of changes

Correcting some things that were quickly noticeable when users submitted new component packages.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by graham33 — APPROVED** [review]
Makes sense, thanks!

**Comment by mweinelt** [comment]
I think this kind of naming scheme does not work with `with`.

```
with home-assistant-custom-components; [
  "mweinelt/prometheus_sensor" # is a string
  mweinelt/prometheus_sensor # is a path
];

File: pkgs/servers/home-assistant/custom-components/README.md [hunk 1]
```diff
@@ -40,12 +40,27 @@ buildHomeAssistantComponent {
...
+The project [mweinelt/ha-prometheus-sensor](https://github.com/mweinelt/ha-prometheus-sensor/blob/1.0.0/custom_components/prometheus_sensor/manifest.json#L2)
+would receive the attribute name `"mweinelt/prometheus_sensor"`,
+because the owner is `mweinelt` and the domain is `prometheus_sensor`.
+
+## Package name
+
+The `pname` attribute is a composition of both `owner` and `domain`.
+
+Exposing the `domain` attribute separately allows checking for
+conflicting components at eval time.
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/servers/home-assistant/custom-components/README.md:63]:
```patch
diff --git a/nixos/modules/services/home-automation/home-assistant.nix b/nixos/modules/services/home-automation/home-assistant.nix
index 54fd3e17292f..748ec43548ed 100644
--- a/nixos/modules/services/home-automation/home-assistant.nix
+++ b/nixos/modules/services/home-automation/home-assistant.nix
@@ -81,6 +81,15 @@ let
       type = "module";
     }) cfg.customLovelaceModules;
   };
+
+  duplicates = list: let
+    s = naturalSort list;
+  in concatMap (x: if x.fst == x.snd then [x.fst] else []) (lib.zipLists s (tail s));
+
+  customComponentDomains = map (x: x.domain) (with pkgs.home-assistant-custom-components; [ mweinelt/prometheus_sensor mweinelt/prometheus_sensor ]);
+   #cfg.customComponents;
+
+  duplicateCustomComponentDomains = duplicates customComponentDomains;
 in {
   imports = [
     # Migrations in NixOS 22.05
@@ -420,6 +429,10 @@ in {
         assertion = cfg.openFirewall -> cfg.config != null;
         message = "openFirewall can only be used with a declarative config";
       }
+      {
+        assertion = duplicateCustomComponentDomains != [];
+        message = "Only one custom component for each domain supported: ${concatMapStringsSep ", " duplicateCustomComponentDomains}";
+      }
     ];
 
     networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ cfg.config.http.server_port ];

File: pkgs/servers/home-assistant/custom-components/README.md [hunk 2]
```diff
@@ -40,12 +40,27 @@ buildHomeAssistantComponent {
...
 
-## Package name normalization
+## Package attribute
 
-Apply the same normalization rules as defined for python packages in
-[PEP503](https://peps.python.org/pep-0503/#normalized-names).
-The name should be lowercased and dots, underlines or multiple
-dashes should all be replaced by a single dash.
+The attribute must be composed of both the repository owner as well as
+the domain name as found in the manifest.
```
Comment (SuperSandro2000) [inline pkgs/servers/home-assistant/custom-components/README.md:46]:
```suggestion
the domain name as found in the manifest.json.
```

File: pkgs/servers/home-assistant/custom-components/default.nix [hunk 3]
```diff
@@ -2,5 +2,5 @@
 }:
 
 {
-  prometheus-sensor = callPackage ./prometheus-sensor {};
+  "mweinelt/prometheus_sensor" = callPackage ./mweinelt/prometheus_sensor {};
```
Comment (SuperSandro2000) [inline pkgs/servers/home-assistant/custom-components/default.nix:5]:
```suggestion
  "mweinelt/prometheus_sensor" = callPackage ./mweinelt-prometheus_sensor {};
```
I don't think we need nested directories here

**Comment by SuperSandro2000** [comment]
> But dashes may be contained in the `owner` name, meh.

We could just ditch that you can see on first look which is the owner and what the domain.

**Comment by mweinelt** [comment]
> > But dashes may be contained in the `owner` name, meh.
> 
> We could just ditch that you can see on first look which is the owner and what the domain.

I don't want to.

**Comment by mweinelt** [comment]
Okay, according to a smol script among 955 domains there are only two domains that have more than one owner.

```
955 domains
midea-ac-py ['andersonshatch', 'mill1000']
hon ['Andre0512', 'gvigroux']
```

For midea-ac-py one is inactive, the other is a fork.
For hon, there are indeed two competing integrations.

Maybe we don't need to be overzealous here.

File: pkgs/servers/home-assistant/custom-components/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/servers/home-assistant/custom-components/README.md:63]: Not important for now.

**Pull request merged**